File: testing/fixtures/database/test-database-operations.template.py
"""
Template for database operations testing (PostgreSQL) in Python
"""
import pytest
import psycopg2
from psycopg2.extras import RealDictCursor


class TestDatabaseOperations:
    """Test PostgreSQL database operations"""

    @pytest.fixture(scope='class')
    def db_connection(self):
        """Setup database connection"""
        conn = psycopg2.connect(
            host=os.environ.get('POSTGRES_HOST', 'localhost'),
            port=os.environ.get('POSTGRES_PORT', 5432),
            database=os.environ.get('POSTGRES_DB', 'cortex_test'),
            user=os.environ.get('POSTGRES_USER', 'cortex_test'),
            password=os.environ.get('POSTGRES_PASSWORD', 'cortex_test_password')
        )

        # Setup test tables
        with conn.cursor() as cur:
            cur.execute("""
                CREATE TABLE IF NOT EXISTS test_items (
                    id SERIAL PRIMARY KEY,
                    name VARCHAR(255) NOT NULL,
                    description TEXT,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                );
            """)
            conn.commit()

        yield conn

        # Cleanup
        with conn.cursor() as cur:
            cur.execute('DROP TABLE IF EXISTS test_items;')
            conn.commit()
        conn.close()

    @pytest.fixture(autouse=True)
    def cleanup_data(self, db_connection):
        """Clean up test data after each test"""
        yield
        with db_connection.cursor() as cur:
            cur.execute('TRUNCATE test_items RESTART IDENTITY CASCADE;')
            db_connection.commit()

    def test_insert_record(self, db_connection):
        """Test inserting a new record"""
        with db_connection.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(
                "INSERT INTO test_items (name, description) VALUES (%s, %s) RETURNING *",
                ('Test Item', 'Test Description')
            )
            result = cur.fetchone()
            db_connection.commit()

        assert result is not None
        assert result['id'] is not None
        assert result['name'] == 'Test Item'
        assert result['description'] == 'Test Description'

    def test_insert_missing_required_field(self, db_connection):
        """Test insert fails with missing required field"""
        with pytest.raises(psycopg2.Error):
            with db_connection.cursor() as cur:
                cur.execute(
                    "INSERT INTO test_items (description) VALUES (%s)",
                    ('No name',)
                )
                db_connection.commit()
        db_connection.rollback()

    def test_insert_multiple_records(self, db_connection):
        """Test inserting multiple records"""
        items = [
            ('Item 1', 'Description 1'),
            ('Item 2', 'Description 2'),
            ('Item 3', 'Description 3')
        ]

        with db_connection.cursor() as cur:
            for name, description in items:
                cur.execute(
                    "INSERT INTO test_items (name, description) VALUES (%s, %s)",
                    (name, description)
                )
            db_connection.commit()

            cur.execute('SELECT COUNT(*) FROM test_items')
            count = cur.fetchone()[0]

        assert count == 3

    def test_select_all_records(self, db_connection):
        """Test retrieving all records"""
        # Insert test data
        with db_connection.cursor() as cur:
            cur.execute(
                "INSERT INTO test_items (name, description) VALUES (%s, %s)",
                ('Test Item', 'Test Description')
            )
            db_connection.commit()

            cur.execute('SELECT * FROM test_items')
            results = cur.fetchall()

        assert len(results) > 0

    def test_select_by_id(self, db_connection):
        """Test retrieving a specific record by id"""
        with db_connection.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(
                "INSERT INTO test_items (name, description) VALUES (%s, %s) RETURNING id",
                ('Specific Item', 'Specific Description')
            )
            item_id = cur.fetchone()['id']
            db_connection.commit()

            cur.execute('SELECT * FROM test_items WHERE id = %s', (item_id,))
            result = cur.fetchone()

        assert result is not None
        assert result['name'] == 'Specific Item'

    def test_select_with_filter(self, db_connection):
        """Test filtering records with WHERE clause"""
        with db_connection.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(
                "INSERT INTO test_items (name, description) VALUES (%s, %s)",
                ('Filtered Item', 'Filtered Description')
            )
            db_connection.commit()

            cur.execute("SELECT * FROM test_items WHERE name LIKE %s", ('%Filtered%',))
            results = cur.fetchall()

        assert len(results) > 0
        assert 'Filtered' in results[0]['name']

    def test_update_record(self, db_connection):
        """Test updating a record"""
        with db_connection.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(
                "INSERT INTO test_items (name, description) VALUES (%s, %s) RETURNING id",
                ('Original Name', 'Original Description')
            )
            item_id = cur.fetchone()['id']
            db_connection.commit()

            cur.execute(
                "UPDATE test_items SET name = %s, updated_at = CURRENT_TIMESTAMP WHERE id = %s",
                ('Updated Name', item_id)
            )
            db_connection.commit()

            cur.execute('SELECT * FROM test_items WHERE id = %s', (item_id,))
            result = cur.fetchone()

        assert result['name'] == 'Updated Name'

    def test_update_nonexistent_record(self, db_connection):
        """Test updating non-existent record returns 0 rows"""
        with db_connection.cursor() as cur:
            cur.execute(
                "UPDATE test_items SET name = %s WHERE id = %s",
                ('Updated Name', 99999)
            )
            db_connection.commit()
            assert cur.rowcount == 0

    def test_delete_record(self, db_connection):
        """Test deleting a record"""
        with db_connection.cursor() as cur:
            cur.execute(
                "INSERT INTO test_items (name, description) VALUES (%s, %s) RETURNING id",
                ('To Be Deleted', 'Will be removed')
            )
            item_id = cur.fetchone()[0]
            db_connection.commit()

            cur.execute('DELETE FROM test_items WHERE id = %s', (item_id,))
            db_connection.commit()

            cur.execute('SELECT * FROM test_items WHERE id = %s', (item_id,))
            result = cur.fetchone()

        assert result is None

    def test_transaction_commit(self, db_connection):
        """Test successful transaction commit"""
        try:
            with db_connection.cursor() as cur:
                cur.execute(
                    "INSERT INTO test_items (name, description) VALUES (%s, %s)",
                    ('Transaction Item 1', 'Desc 1')
                )
                cur.execute(
                    "INSERT INTO test_items (name, description) VALUES (%s, %s)",
                    ('Transaction Item 2', 'Desc 2')
                )
                db_connection.commit()

                cur.execute('SELECT COUNT(*) FROM test_items')
                count = cur.fetchone()[0]

            assert count == 2
        except Exception as e:
            db_connection.rollback()
            raise e

    def test_transaction_rollback(self, db_connection):
        """Test transaction rollback on error"""
        try:
            with db_connection.cursor() as cur:
                cur.execute(
                    "INSERT INTO test_items (name, description) VALUES (%s, %s)",
                    ('Transaction Item', 'Desc')
                )

                # This should fail (missing required field)
                cur.execute(
                    "INSERT INTO test_items (description) VALUES (%s)",
                    ('Should fail',)
                )
                db_connection.commit()
        except psycopg2.Error:
            db_connection.rollback()

            # Verify rollback
            with db_connection.cursor() as cur:
                cur.execute('SELECT COUNT(*) FROM test_items')
                count = cur.fetchone()[0]
            assert count == 0

File: testing/fixtures/database/test-database-operations.template.js
// Template for database operations testing (PostgreSQL)

const { Client } = require('pg');

describe('Database Operations', () => {
  let client;

  beforeAll(async () => {
    // Setup database connection
    client = new Client({
      host: process.env.POSTGRES_HOST || 'localhost',
      port: process.env.POSTGRES_PORT || 5432,
      database: process.env.POSTGRES_DB || 'cortex_test',
      user: process.env.POSTGRES_USER || 'cortex_test',
      password: process.env.POSTGRES_PASSWORD || 'cortex_test_password'
    });

    await client.connect();

    // Setup test tables
    await client.query(`
      CREATE TABLE IF NOT EXISTS test_items (
        id SERIAL PRIMARY KEY,
        name VARCHAR(255) NOT NULL,
        description TEXT,
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
        updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      );
    `);
  });

  afterAll(async () => {
    // Cleanup test tables
    await client.query('DROP TABLE IF EXISTS test_items;');
    await client.end();
  });

  afterEach(async () => {
    // Clean up test data after each test
    await client.query('TRUNCATE test_items RESTART IDENTITY CASCADE;');
  });

  describe('INSERT operations', () => {
    it('should insert a new record', async () => {
      const result = await client.query(
        'INSERT INTO test_items (name, description) VALUES ($1, $2) RETURNING *',
        ['Test Item', 'Test Description']
      );

      expect(result.rows).toHaveLength(1);
      expect(result.rows[0]).toHaveProperty('id');
      expect(result.rows[0].name).toBe('Test Item');
    });

    it('should fail to insert with missing required fields', async () => {
      await expect(
        client.query('INSERT INTO test_items (description) VALUES ($1)', ['No name'])
      ).rejects.toThrow();
    });

    it('should insert multiple records', async () => {
      const items = [
        ['Item 1', 'Description 1'],
        ['Item 2', 'Description 2'],
        ['Item 3', 'Description 3']
      ];

      for (const [name, description] of items) {
        await client.query(
          'INSERT INTO test_items (name, description) VALUES ($1, $2)',
          [name, description]
        );
      }

      const result = await client.query('SELECT COUNT(*) FROM test_items');
      expect(parseInt(result.rows[0].count)).toBe(3);
    });
  });

  describe('SELECT operations', () => {
    beforeEach(async () => {
      // Insert test data
      await client.query(
        'INSERT INTO test_items (name, description) VALUES ($1, $2)',
        ['Test Item', 'Test Description']
      );
    });

    it('should retrieve all records', async () => {
      const result = await client.query('SELECT * FROM test_items');
      expect(result.rows.length).toBeGreaterThan(0);
    });

    it('should retrieve a specific record by id', async () => {
      const insertResult = await client.query(
        'INSERT INTO test_items (name, description) VALUES ($1, $2) RETURNING id',
        ['Specific Item', 'Specific Description']
      );
      const id = insertResult.rows[0].id;

      const result = await client.query(
        'SELECT * FROM test_items WHERE id = $1',
        [id]
      );

      expect(result.rows).toHaveLength(1);
      expect(result.rows[0].name).toBe('Specific Item');
    });

    it('should filter records with WHERE clause', async () => {
      await client.query(
        'INSERT INTO test_items (name, description) VALUES ($1, $2)',
        ['Filtered Item', 'Filtered Description']
      );

      const result = await client.query(
        'SELECT * FROM test_items WHERE name LIKE $1',
        ['%Filtered%']
      );

      expect(result.rows.length).toBeGreaterThan(0);
      expect(result.rows[0].name).toContain('Filtered');
    });
  });

  describe('UPDATE operations', () => {
    let testItemId;

    beforeEach(async () => {
      const result = await client.query(
        'INSERT INTO test_items (name, description) VALUES ($1, $2) RETURNING id',
        ['Original Name', 'Original Description']
      );
      testItemId = result.rows[0].id;
    });

    it('should update a record', async () => {
      await client.query(
        'UPDATE test_items SET name = $1, updated_at = CURRENT_TIMESTAMP WHERE id = $2',
        ['Updated Name', testItemId]
      );

      const result = await client.query(
        'SELECT * FROM test_items WHERE id = $1',
        [testItemId]
      );

      expect(result.rows[0].name).toBe('Updated Name');
    });

    it('should return 0 rows affected for non-existent record', async () => {
      const result = await client.query(
        'UPDATE test_items SET name = $1 WHERE id = $2',
        ['Updated Name', 99999]
      );

      expect(result.rowCount).toBe(0);
    });
  });

  describe('DELETE operations', () => {
    let testItemId;

    beforeEach(async () => {
      const result = await client.query(
        'INSERT INTO test_items (name, description) VALUES ($1, $2) RETURNING id',
        ['To Be Deleted', 'Will be removed']
      );
      testItemId = result.rows[0].id;
    });

    it('should delete a record', async () => {
      await client.query('DELETE FROM test_items WHERE id = $1', [testItemId]);

      const result = await client.query(
        'SELECT * FROM test_items WHERE id = $1',
        [testItemId]
      );

      expect(result.rows).toHaveLength(0);
    });

    it('should delete multiple records', async () => {
      await client.query(
        'INSERT INTO test_items (name, description) VALUES ($1, $2), ($3, $4)',
        ['Item 1', 'Desc 1', 'Item 2', 'Desc 2']
      );

      await client.query('DELETE FROM test_items WHERE name LIKE $1', ['Item%']);

      const result = await client.query(
        'SELECT * FROM test_items WHERE name LIKE $1',
        ['Item%']
      );

      expect(result.rows).toHaveLength(0);
    });
  });

  describe('Transaction operations', () => {
    it('should commit successful transaction', async () => {
      await client.query('BEGIN');

      try {
        await client.query(
          'INSERT INTO test_items (name, description) VALUES ($1, $2)',
          ['Transaction Item 1', 'Desc 1']
        );
        await client.query(
          'INSERT INTO test_items (name, description) VALUES ($1, $2)',
          ['Transaction Item 2', 'Desc 2']
        );

        await client.query('COMMIT');

        const result = await client.query('SELECT COUNT(*) FROM test_items');
        expect(parseInt(result.rows[0].count)).toBe(2);
      } catch (error) {
        await client.query('ROLLBACK');
        throw error;
      }
    });

    it('should rollback failed transaction', async () => {
      await client.query('BEGIN');

      try {
        await client.query(
          'INSERT INTO test_items (name, description) VALUES ($1, $2)',
          ['Transaction Item', 'Desc']
        );

        // This should fail (no name column value)
        await client.query(
          'INSERT INTO test_items (description) VALUES ($1)',
          ['Should fail']
        );

        await client.query('COMMIT');
      } catch (error) {
        await client.query('ROLLBACK');

        // Verify rollback
        const result = await client.query('SELECT COUNT(*) FROM test_items');
        expect(parseInt(result.rows[0].count)).toBe(0);
      }
    });
  });
});
